### ISSUE TITLE
OSHI5 --> OSHI-future

### ISSUE THREAD

**Issue by dbwiddis** [oshi/oshi5#16]
I'm finding a lot more API imporovements/adjustments that I want to make, and foresee an OSHI 5 release long before I expect to see this new project released.  I wanted to give a heads up that we might not want to put a specific number on this future redesign/release.

**Comment by cilki** [comment]
No problem. Are you still behind the idea of a single-source-of-truth JSON file defining all attributes from which an API layer can be automatically generated and organizing the fetching layer into stateless "query methods" that are dynamically invoked by OSHI at runtime?

Those are the two fundamental principles in my design effort. Although the current branch is brittle, it's already working for a few simple attributes on Linux. Adding more features like CPU would increase my confidence that the model is general enough to work for the whole project.

**Comment by cilki** [comment]
Also, I forgot to post [this paper](https://github.com/cilki/CS370/blob/master/Final/Final.pdf) that I wrote for a class which may be interesting or inspiring.

**Comment by dbwiddis** [comment]
Yes, I think auto-generation is a great idea for standardizing code and avoiding repetition.  My thought is:
- generated API which defines the `queryX()` methods required.
- `queryX()` methods are the middle layer that implements logic (e.g., one-time fetch or update values, using backup methods when primary failure, grabbing multiple attributes at once, using memoizers for thread safe caching, etc.) all of which call "driver" methods
- driver methods that implement the actual fetching of data.  Some are specific (run this command line and parse) and some are generic with parameters (e.g., WMI queries)

**Comment by cilki** [comment]
Sounds good :+1:. I'll continue working on this in my spare time.

**Comment by dbwiddis** [comment]
Closing this issue as I've already renamed this to oshi6...hopefully that's the final name but in case it has to move to oshi7 at some point, don't be surprised! :) 

A few notes in the context of my remarks above:
 - I've been trying to move the API in the direction cited above, creating (stateless) driver classes and modeling three-layer structure envisioned (cross-platform API / OS-specific logic calling drivers (sometimes with memoizer, and wrapping or copying result for thread safety) with fallbacks / drivers handling the specifics).  Obviously it's a big task to move everything so I'm not trying to refactor working code, but I have done mass movements of types of code (perf counters, WMI, proc) and am moving that way with any new additions.
 - In my exploration of modules and restricting APIs, I've found there are a few cases where in order to answer a user request for a platform-specific feature, I've made it available in a "driver" class without exposing it to the API.  I actually think this is a reasonable compromise and think that in a modular system we should expose the drivers to users.
 - The package names for oshi v1-v5 is "oshi.*".  For modules I'd want to rewrite the packages as `com.github.oshi.*`.  I'm toying with the idea of both structures living alongside each other, possibly in a multi-release JAR.
